### PR TITLE
[Remove dependencies] Convert some default format to external package

### PR DESCRIPTION
**Pull request by A-312**
fix: #321 
fix: #317

validator and moment is no longer a dependencie of convict

**WARNING :**

Now, you should publish with `lerna publish` instead of `npm publish`

### CODE REVIEW

**Comment by A-312** [comment]
Ready ! 😄

**Comment by A-312** [comment]
@madarche Can you watch for merge this PR in the next week ? I would like [PR my other issues (3)](https://github.com/mozilla/node-convict/issues/created_by/A-312).

**Comment by A-312** [comment]
Oh, Coverage decrease. :( I will fix it.

 - [this code is not tested](https://coveralls.io/builds/25999754/source?filename=packages/convict-validator/lib/index.js#L11) ;
 - [The instruction for exclude `test` folder is wrong](https://github.com/A-312/node-convict/blob/nodep/nycrc.json#L5).

**Comment by A-312** [comment]
?

**Comment by madarche** [comment]
@A-312 I've been over-busy those last weeks, sorry! I'll deal with everything related to this package this weekend.

**Comment by A-312** [comment]
Just know that PR will be look it's ok for me.

**Comment by A-312** [comment]
I check my rebase/fix conflicts seems to be good.

**Comment by madarche** [comment]
@A-312 in this PR there are many many files that have been changed (135). Obviously many shouldn't have been. Probably problems happened on your rebase. Maybe could you create a totally new and clean PR based on current `master`? Thank you.

**Comment by A-312** [comment]
> @A-312 in this PR there are many many files that have been changed (135). Obviously many shouldn't have been. Probably problems happened on your rebase. Maybe could you create a totally new and clean PR based on current `master`? Thank you.

This is my PR, see : 

 - https://github.com/A-312/node-convict/blob/nodep/README.md
 - https://github.com/A-312/node-convict/tree/nodep/packages

**Comment by A-312** [comment]
@madarche Now with lerna, this repos have 3 packages

**Comment by A-312** [comment]
Next week end ? 😕

**Comment by A-312** [comment]
135 files change because : (I will try to count files changed)

 - First commit, I move config files (9), `./lib/` and `/test/` to `./packages/convict/`. *test files* = ~90 files ;
 - After, I made package : `convict-moment` and `convict-moment` with their own test, readme, config and js file (14 files = 7 files x 2) ;

= 113 files

 - Edits exemples (2 files), readme, config (9 files).

= 124 files

124 is near of 135, that seem good.

**Comment by A-312** [comment]
Do you plan to review this PR ? I made several commits to help you to review.

**Comment by A-312** [comment]
@madarche Do you plan to review this PR ? I made several commits to help you to review.

**Comment by A-312** [comment]
Maybe someone else can be help to review my PR.

**Comment by A-312** [comment]
Ping @madarche

**Comment by A-312** [comment]
Can I get a date or something like "one month", "in January/feb" ?

**Comment by A-312** [comment]
@vladikoff Do you have news about @madarche ? No activity since 2 months :( I Just need to know if I need to make my change on my repos or here.

**Comment by A-312** [comment]
@mozilla Sorry, but i don't have any news of a maintener of this repository (for my several PR since 2 months).

**Comment by vladikoff** [comment]
I shall take a look at this during the week

**Comment by madarche** [comment]
Hello all, I'm sorry for the absence. I've been trying hard to to devote some time to convict all this time. I'm back for as much as I can.

**Comment by madarche** [comment]
First, I'm all for the proposed change: This is a very good idea and goes in the right direction for making convict simpler and easier to maintain and contribute to. And we should release convict@6 when this PR is merged.

 (The first time I read this PR, *lerna* was not mentioned in the description and so I thought it was not about multi-packages, and I was too much in a hurry. Anyway it's all very clear now.)

Here are some first remarks:

* Could you rebase this PR, because it doesn't include at least some parts of the commits c90e7e991040364842a62811cbe4a6787993ad16 and 0cbc321243aab6a95951c0707847971f981929d7

* The `convict-validator` package name may not be very well chosen, as it makes think that the *validation* functionality of convict is provided by this package. What would you think about the name `convict-email-ipaddress-url` instead? Yest it's long but it's all very clear.

I'll update this comment with other remarks. Those first are here to let you know that I care and I've started the review work.

**Comment by A-312** [comment]
CHECKLIST : in progress.
> 
> * [x] I have created a new branch `feat-multi-packages-split`. Could you rework this PR to make it against this branch please? That way we could work together in the branch.
> * [ ] Could rework this PR, because it doesn't include at least some parts of the commits [c90e7e9](https://github.com/mozilla/node-convict/commit/c90e7e991040364842a62811cbe4a6787993ad16) and [0cbc321](https://github.com/mozilla/node-convict/commit/0cbc321243aab6a95951c0707847971f981929d7)
> * [ ] Don't copy/put `.eslintrc` and `.eslintignore` files in every packages, as all the eslint scripts will use the files at the root just fine.
> * [ ] The `convict-validator` package name may not be very well chosen, as it makes think that the _validation_ functionality of convict is provided by this package. What would you think about the name `convict-format-email-ipaddress-url`. Yes it's long but it's all very clear. Likewise `convict-moment` should then be renamed into `convict-format-duration-timestamp`. What do you think?

**Comment by A-312** [comment]
> Don't copy/put .eslintrc and .eslintignore files in every packages, as all the eslint scripts will use the files at the root just fine.

These files will be missing if we download the package with npm and try to run lint but I think is not usefull to run eslint in node_module instead of github folder.

**Comment by madarche** [comment]
> These files will be missing if we download the package with npm and try to run lint but I think is not usefull to run eslint in node_module instead of github folder.

You are right, those files will be missing if those packages are downloaded from the NPM registry to be tested! At the moment, we, developers of Convict are always working from the Git repository, but that might be useful for someone evaluating the convict package from `node_modules`, and we can dream, in the future the NPM registry could automatically launch tests on a publish request to better assert the security of a package.

So I withdraw this remark. Sorry! And again thanks for the good insight.

**Comment by A-312** [comment]
> The `convict-validator` package name may not be very well chosen, as it makes think that the _validation_ functionality of convict is provided by this package. What would you think about the name `convict-format-email-ipaddress-url`. Yes it's long but it's all very clear. Likewise `convict-moment` should then be renamed into `convict-format-duration-timestamp`. What do you think?

I agree with that `convict-validator` and `convict-format` was temporary. But what do you think of : 

 - "convict-format-with-moment"
 - "convict-format-with-validator" or "convict-format-validator-support"

I don't think if we have several package in one we should add each format name.

Because some people can extend/add format with validator function. Or maybe do something dynamically like that : 

```js
convict.addFormats(require("convict-format-with-validator")({
   "currency": "isCurrency",
   "email": "isEmail"
}));
// or : 
convict.addFormats(require("convict-format-with-validator")([ "currency", "email" ]));
// a function will add first uppercase and "is" to convert the format as validator function "currency" => "isCurrency"
```

**Comment by madarche** [comment]
I'm 100% convinced! 😀

As for the names, I like `convict-format-with-moment` and `convict-format-with-validator` the most.

**Comment by A-312** [comment]
@madarche done :D

File: package.json [hunk 1]
```diff
@@ -1,56 +1,36 @@
...
-    "version": "./assert_changelog_ready $npm_package_version"
-  },
-  "files": [
-    "lib"
-  ],
-  "main": "lib/convict.js",
+  "name": "node-convict",
+  "private": true,
   "dependencies": {
+    "json5": "2.1.0",
```
Comment (madarche) [inline package.json]: This `json5` should not be present anymore as a `dependency`

File: package.json [hunk 2]
```diff
@@ -1,56 +1,36 @@
...
     "lodash.clonedeep": "4.5.0",
     "moment": "2.24.0",
-    "validator": "11.1.0",
+    "validator": "10.11.0",
```
Comment (madarche) [inline package.json]: It's `validator@11.1.0` in `master`

File: package.json (hunk 1 above)
Comment (A-312) [reply package.json]: Now, json is devDepencies.

ok

File: package.json (hunk 2 above)
Comment (A-312) [reply package.json]: done

**Comment by A-312** [comment]
@madarche done

**Comment by madarche** [comment]
@A-312 this is all very good, thanks! So I'm merging your work into `feat-multi-packages-split`.

In the next following days I'll be doing some very little edits in this branch and I'll be using this branch on some projects, before finally merging into `master`. That will be convict@6 then 😁

**Pull request merged**